### PR TITLE
W2V modifications for graph generation

### PR DESCRIPTION
**Pull request by MissingNO57**

### CODE REVIEW

File: libs/ml/examples/word2vec_distributed_learning/main.cpp [hunk 1]
```diff
@@ -129,6 +130,7 @@ int main(int argc, char **argv)
     cp.data                        = {client_data[i]};
     auto client =
         std::make_shared<Word2VecClient<TensorType>>(std::to_string(i), cp, console_mutex_ptr);
+    client->SetMaxUpdates(iterations_count);
     // TODO(1597): Replace ID with something more sensible
```
Comment (evsmithx) [inline libs/ml/examples/word2vec_distributed_learning/main.cpp]: This comment isn't in the right place any more! I think we decided not to do anything about this issue anyway, so maybe we can remove it?

File: libs/ml/examples/word2vec_distributed_learning/main.cpp [hunk 2]
```diff
@@ -172,6 +174,24 @@ int main(int argc, char **argv)
       t.join();
     }
 
+    std::ofstream lossfile("losses.csv", std::ofstream::out | std::ofstream::app);
```
Comment (evsmithx) [inline libs/ml/examples/word2vec_distributed_learning/main.cpp]: It would be good if this was a commandline arg.

**Review by evsmithx — CHANGES_REQUESTED** [review]

**Review by evsmithx — DISMISSED** [review]

File: libs/ml/examples/word2vec_SGNS_dataloader/main.cpp [hunk 3]
```diff
@@ -16,6 +16,7 @@
 //
 //------------------------------------------------------------------------------
 
+#include "../../src/utilities/word2vec_utilities.cpp"
```
Comment (evsmithx) [inline libs/ml/examples/word2vec_SGNS_dataloader/main.cpp]: Shouldn't you still be including the .hpp file not the .cpp?

File: libs/ml/examples/word2vec_distributed_learning/word2vec_client.hpp [hunk 4]
```diff
@@ -57,8 +57,27 @@ class Word2VecClient : public TrainingClient<TensorType>
...
+  DataType GetAnalogyScore()
+  {
+    // first get hold of the skipgram layer by searching the return name in the graph
+    std::shared_ptr<fetch::ml::layers::SkipGram<TensorType>> sg_layer =
+        std::dynamic_pointer_cast<fetch::ml::layers::SkipGram<TensorType>>(
+            (this->g_ptr_->GetNode(skipgram_))->GetOp());
+
+    // next get hold of the embeddings
+    std::shared_ptr<fetch::ml::ops::Embeddings<TensorType>> embeddings =
+        sg_layer->GetEmbeddings(sg_layer);
```
Comment (evsmithx) [inline libs/ml/examples/word2vec_distributed_learning/word2vec_client.hpp]: There's now utilities::GetEmbeddings for this

File: libs/ml/examples/word2vec_graph_evaluation/main.cpp [hunk 5]
```diff
@@ -16,13 +16,13 @@
 //
 //------------------------------------------------------------------------------
 
+#include "../../src/utilities/word2vec_utilities.cpp"
```
Comment (evsmithx) [inline libs/ml/examples/word2vec_graph_evaluation/main.cpp]: Same here, include the .hpp file

File: libs/ml/include/ml/distributed_learning/distributed_learning_client.hpp [hunk 6]
```diff
@@ -17,6 +17,7 @@
 //
 //------------------------------------------------------------------------------
 
+#include "../../../src/utilities/word2vec_utilities.cpp"
```
Comment (evsmithx) [inline libs/ml/include/ml/distributed_learning/distributed_learning_client.hpp]: hpp

File: libs/ml/src/utilities/word2vec_utilities.cpp [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+//------------------------------------------------------------------------------
+
+#include "ml/utilities/word2vec_utilities.hpp"
+
+namespace fetch {
+namespace ml {
+namespace utilities {
+
+// Timestamp for logging
+std::string GetStrTimestamp()
```
Comment (evsmithx) [inline libs/ml/src/utilities/word2vec_utilities.cpp:26]: This should be declared in the header file

**Review by evsmithx — CHANGES_REQUESTED** [review]

File: libs/ml/src/utilities/word2vec_utilities.cpp (hunk 7 above)
Comment (MissingNO57) [reply libs/ml/src/utilities/word2vec_utilities.cpp:26]: It can't be, because it is not templated so it wouldn't pass static analysis.

**Review by evsmithx — APPROVED** [review]

**Pull request merged**